### ISSUE TITLE
Make monadic transformation

### ISSUE THREAD

**Issue by clarus** [formal-land/coq-of-rust#60]
A monadic transformation is necessary to represent the various side-effects, like control-flow primitives (`return`, `break`, `continue`), errors (`panic!`), and state mutations.

There is already a tentative pull-request to make a monadic transformation of the generated Coq code: https://github.com/formal-land/coq-of-rust/pull/58

This pull-request avoids making an explicit monadic transform through the use of tactics. But the problem is that it limits type inference as we do not have access to the expected return type of an expression inside the tactic. Type inference is quite fragile with what we are doing, as we already rely on type-classes that need a lot inference for example.

A more reliable solution is to make an **explicit monadic transformation**. This is the goal of this task.

**Comment by clarus** [comment]
Here is an example of expected transformation (many variations are possible):

From the Rust code:
```rust
fn foo(n: i32) -> i32 {
  return n + 1;
}

fn bar(m: i32) -> i32 {
  return foo(foo(m));
}
```
generate for the `bar` function the code:
```coq
Definition bar (m : i32) : M i32 :=
  let* v1 := foo m in
  let* v2 := foo v1 in
  M.Pure v2.
```

**Comment by andreykl** [comment]
I didn't reach to much.
I've checked a bit what is in another branches (and will examining it in deeper details now).
So, just in order to discuss/share the ideas.

Here is a variant of monadic transformation for mutable state.
https://github.com/formal-land/coq-of-rust/blob/ak%40towards-monadic-transformation/coq_translation/examples-from-rust-book/freer.v#L621

I would like to put example of code below. This manually translated from here 
https://github.com/formal-land/coq-of-rust/blob/67ac46b4894cc7fa272348f1ca65993b47823f69/examples-from-rust-book/functions/associated_functions_and_methods.rs#L55

```coq
Definition translate (x : f64) (y : f64) : EffState Self unit :=
    self <- getEff ;;
    putEff (self.["p1"] :=u (Point.set_x (fun v => v + x))) ;;
    self <- getEff ;;
    putEff (self.["p2"] :=u (Point.set_x (fun v => v + x))) ;;
    self <- getEff ;;
    putEff (Rectangle.set_p1 (Point.set_y (fun v => v + y)) self) ;;
    self <- getEff ;;
    putEff (Rectangle.set_p2 (Point.set_y (fun v => v + y)) self).

  Global Instance Method_translate : Notation.Dot "translate" := {
      Notation.dot := translate;
    }.
```
Code above compiles (inside freer.v)
So, the idea is to put all mutable variables into state one by one. So, for rust function with signature
`fn translate(&mut self, &mut x: f64, &mut y: f64) {` we should have definition like

```coq 
Definition translate : EffState (Self, f64, f64) unit :=
    (self, x, y) <- getEff ;;
    putEff (self.["p1"] :=u ((Point.set_x (fun v => v + x)), x, y)) ;;
    (self, x, y) <- getEff ;;
```
etc..

One of problems with this code - I'm not able to write `rect.["p1"].["x"] :=u 44`. I have to write `rect.["p1"] :=u Point.set_x 44`
I found this https://github.com/tchajed/coq-record-update , it might be possible to use some ideas from them to make chainable update syntax (i'm not sure).

Another problem is that we need handle somehow chained functions calls. (so, call one function with mutable arguments inside another one. Problem that we (should?) have 2 variables named `self`, so, we will need maintain names, like self1, self2.. etc)

**Comment by dhilst** [comment]
> One of problems with this code - I'm not able to write `rect.["p1"].["x"] :=u 44`. I have to write `rect.["p1"] :=u Point.set_x 44`

I have this problem in coq-of-solidity, I have a nested mapping update syntax there, which we may tweak to look more like rust

https://gitlab.com/formal-land/coq-of-solidity/-/blob/master/proofs/Prelude/Mapping.v

**Comment by dhilst** [comment]
Oh in fact this is record update syntax right? Not mapping, sorry. The syntax may be useful yet. I do foo[ a, b, c ] instead of foo[a][b][c], the write a function that receives a path (a list of indexes) instead of a single index

**Comment by andreykl** [comment]
Yes, it looks like your solution is at least very close to what I would like to have!
I think they are similar (I mean mapping and update syntax). Yes, I think we might need to use syntax `foo[ a, b, c ]`. I'm not sure yet what does Guillaume mean with his `(| .. |)` syntax.

**Comment by dhilst** [comment]
`g(| ... |)` should be a notation for function application for now, I'm working on it #62
